gan/trainer.py

```diff
@@ -94,12 +104,59 @@ def train_epoch(self, loader):
                 self.g_optimizer.step()
                 k_it = 0
                 n_g_steps += 1
+                del G_loss
+
+            del sample, G_sample, D_real, D_logit_real, D_loss, D_fake, D_logit_fake
 
         if self.visualizer is not None:
-            self.visualizer.update_losses(g_loss=G_train_loss / n_g_steps, d_loss=D_train_loss / n_d_steps,
+            self.visualizer.update_losses(epoch=self.current_epoch,
+                                          g_loss=G_train_loss / n_g_steps,
+                                          d_loss=D_train_loss / n_d_steps,
                                           type='train')
             self.visualizer.show_generator_results(generator=self.generator)
 
+    def valid_epoch(self, loader):
+        self.generator.eval()
+        self.discriminator.eval()
+        G_valid_loss = 0.0
+        D_valid_loss = 0.0
+        n_steps = 0
+        generator_loss = GeneratorLoss()
+        discriminator_loss = DiscriminatorLoss(label_smoothing=self.config.label_smoothing)
+        for sample in loader:
+            D_real, D_logit_real = self.discriminator(*sample)
+            noise = self.noise_sampler.sample_batch(loader.batch_size)
+            # empty tuple for not conditional gan
+            condition = noise[1:]
+
+            # G_sample is batch of images generated by generator using noise as input
+            G_sample = self.generator(*noise)
+            D_fake, D_logit_fake = self.discriminator(G_sample, *condition)
+
+            D_loss = discriminator_loss(D_logit_real, D_logit_fake)
+            D_valid_loss += D_loss.data.cpu().numpy()
+
+            noise = self.noise_sampler.sample_batch(loader.batch_size)
+            # empty tuple for not conditional gan
+            condition = noise[1:]
+            G_sample = self.generator(*noise)
+            D_fake, D_logit_fake = self.discriminator(G_sample, *condition)
+            G_loss = generator_loss(D_logit_fake)
+            G_valid_loss += G_loss.data.cpu().numpy()
+            n_steps += 1
+            del sample, G_sample, D_real, D_logit_real, D_loss, D_fake, D_logit_fake, G_loss
```

Comment:
Is that necessary?

Reply:
Maybe wrapping into with statement will look nicer.

Reply:
Yes, that reduce memory consumption on GPU